core/src/saros/net/xmpp/contact/XMPPContactsService.java

```diff
@@ -0,0 +1,409 @@
+
+      if (contact.setNickname(entry.getName()))
+        notifyListeners(contact, UpdateType.NICKNAME_CHANGED);
+    }
+  }
+
+  private void contactChangedPresence(Presence presence) {
+    if (roster == null) return;
+
+    String address = presence.getFrom();
+    XMPPContact contact = contacts.get(address);
+    if (contact == null) {
+      log.error("Contact " + address + " for presence update not found!");
+      return;
+    }
+
+    JID fqJid = new JID(address);
+    if (presence.isAvailable()) {
+      handleContactResourceAvailable(contact, fqJid, presence);
+    } else {
+      handleContactResourceUnavailable(contact, fqJid);
+    }
+  }
+
+  private void contactCanceledSubscription(JID jid) {
+    XMPPContact contact = contacts.get(jid.getRAW());
+    if (contact == null) {
+      log.debug("Strange behaviour: no contact found for canceled subscription from " + jid);
+      return;
+    }
+
+    contact.removeResources();
+    contact.setBaseStatus(ContactStatus.TYPE_SUBSCRIPTION_CANCELED);
+    notifyListeners(contact, UpdateType.STATUS);
+  }
+
+  private boolean handlePendingSubscription(XMPPContact contact, RosterEntry entry) {
+    if (entry.getStatus() == ItemStatus.SUBSCRIPTION_PENDING)
+      return contact.setBaseStatus(ContactStatus.TYPE_SUBSCRIPTION_PENDING);
+    return false;
+  }
+
+  private void handleContactResourceAvailable(XMPPContact contact, JID fqJid, Presence presence) {
+    contact.setSubscribed();
+
+    ContactStatus status;
+    if (presence.isAway()) {
+      status = ContactStatus.instance(Type.AWAY, presence.getMode().toString());
+    } else {
+      status = ContactStatus.TYPE_AVAILABLE;
+    }
+    if (contact.setResourceStatus(fqJid, status)) notifyListeners(contact, UpdateType.STATUS);
+
+    discoveryService.querySarosSupport(
+        fqJid.getRAW(), handleQuerySarosSupportResult(fqJid, contact));
+  }
+
+  private void handleContactResourceUnavailable(XMPPContact contact, JID fqJid) {
+    discoveryService.removePresence(fqJid.getRAW());
+    if (contact.removeResource(fqJid)) notifyListeners(contact, UpdateType.STATUS);
+  }
+
+  /**
+   * Function to be called after a positive query.
+   *
+   * @param fqJid
+   * @param contact
+   * @return Function to be called after a positive query
+   */
+  private Function<Boolean, Void> handleQuerySarosSupportResult(JID fqJid, XMPPContact contact) {
+    return hasSupport -> {
+      contactsExecutor.execute(
+          () -> {
+            if (contact.updateSarosSupport(fqJid, hasSupport))
+              notifyListeners(contact, UpdateType.SAROS_SUPPORT);
+          });
+      return null;
+    };
+  }
+
+  private void notifyListeners(XMPPContact contact, UpdateType updateType) {
+    log.debug("updated " + contact + " - " + updateType);
+    Optional<XMPPContact> optionalContact = Optional.ofNullable(contact);
+    updateListeners.forEach(listener -> listener.update(optionalContact, updateType));
```

Comment:
for (listener : updateListeners) {try { listener.update(...)}catch(RuntimeException e){...}}

Common pattern for all notifications as you call alien stuff.